state/mongodb/mongodb.go

```diff
@@ -196,6 +216,53 @@ func (m *MongoDB) BulkDelete(req []state.DeleteRequest) error {
 	return nil
 }
 
+// Multi performs a transactional operation. succeeds only if all operations succeed, and fails if one or more operations fail
+func (m *MongoDB) Multi(operations []state.TransactionalRequest) error {
+	sess, err := m.client.StartSession()
+	err = sess.StartTransaction(options.Transaction().
+		SetReadConcern(readconcern.Snapshot()).
+		SetWriteConcern(writeconcern.New(writeconcern.WMajority())),
+	)
+
+	defer sess.EndSession(context.Background())
+
+	if err != nil {
+		return fmt.Errorf("error in starting the transaction: %s", err)
+	}
+
+	sess.WithTransaction(context.Background(), func(sessCtx mongo.SessionContext) (interface{}, error) {
+		err = m.doTransaction(sessCtx, operations)
+		return nil, err
+	})
+
+	return err
+}
+
+func (m *MongoDB) doTransaction(sessCtx mongo.SessionContext, operations []state.TransactionalRequest) error {
+	for _, o := range operations {
+		if o.Operation == state.Upsert {
+			req := o.Request.(state.SetRequest)
+
+			err := m.setInternal(sessCtx, &req)
+
+			if err != nil {
+				sessCtx.AbortTransaction(sessCtx)
+				return fmt.Errorf("error during transaction, aborting the transaction: %s", err)
+			}
+		} else if o.Operation == state.Delete {
+			req := o.Request.(state.DeleteRequest)
+			err := m.deleteInternal(sessCtx, &req)
+
+			if err != nil {
+				sessCtx.AbortTransaction(sessCtx)
+				return fmt.Errorf("error during transaction, aborting the transaction: %s", err)
+			}
```

Comment:
nit: this block is duplicated with line 248-251. I think we can check the error outside of `if-statement`.